File: examples/delphi/fmx/Unit1.pas
{
  Copyright 2022-2022 Michalis Kamburelis.

  This file is part of "Castle Game Engine".

  "Castle Game Engine" is free software; see the file COPYING.txt,
  included in this distribution, for details about the copyright.

  "Castle Game Engine" is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

  ----------------------------------------------------------------------------
}

{ Form unit. }
unit Unit1;

interface

uses
  System.SysUtils, System.Types, System.UITypes, System.Classes, System.Variants,
  FMX.Types, FMX.Controls, FMX.Forms, FMX.Graphics, FMX.Dialogs,
  FMX.Controls.Presentation, FMX.StdCtrls, FMX.Memo.Types, FMX.ScrollBox,
  FMX.Memo,
  Fmx.CastleControl,
  CastleGLUtils, CastleControls, CastleUIControls;

type
  TTestCgeControl = class(TForm)
    Timer1: TTimer;
    PanelSideBar: TPanel;
    Memo1: TMemo;
    Button2D: TButton;
    ButtonUI: TButton;
    Button3D: TButton;
    LabelFps: TLabel;
    CastleControl: TCastleControl;
    procedure FormCreate(Sender: TObject);
    procedure Timer1Timer(Sender: TObject);
    procedure Button3DClick(Sender: TObject);
    procedure Button2DClick(Sender: TObject);
    procedure ButtonUIClick(Sender: TObject);
  public
    { Public declarations }
  end;

var
  TestCgeControl: TTestCgeControl;

implementation

{$R *.fmx}

uses
  CastleRenderOptions, CastleRectangles, CastleColors, CastleRenderContext,
  CastleVectors;

{ TUiTest -------------------------------------------------------------- }

type
  TUiTest = class(TCastleUserInterface)
    constructor Create(Owner: TComponent); override;
    procedure Render; override;
    procedure GLContextOpen; override;
  end;

constructor TUiTest.Create(Owner: TComponent);
begin
  inherited;
  // keep in front, to not be obscured by designs we load using CastleControl.Container.DesignUrl
  KeepInFront := true;
end;

procedure TUiTest.GLContextOpen;
begin
  inherited;
  TestCgeControl.Memo1.Lines.Add(GLInformationString);
end;

procedure TUiTest.Render;
begin
  inherited;
  DrawRectangle(FloatRectangle(5, 5, 10, 10), Yellow);
  FallbackFont.Print(30, 5, Green, FormatDateTime('yyyy-mm-dd, hh:nn:ss', Now));
end;

{ TTestCgeControl ------------------------------------------------------------ }

procedure TTestCgeControl.Button2DClick(Sender: TObject);
begin
  CastleControl.Container.DesignUrl := 'castle-data:/test_2d.castle-user-interface';
end;

procedure TTestCgeControl.Button3DClick(Sender: TObject);
begin
  CastleControl.Container.DesignUrl := 'castle-data:/test_3d.castle-user-interface';
end;

procedure TTestCgeControl.ButtonUIClick(Sender: TObject);
begin
  CastleControl.Container.DesignUrl := 'castle-data:/test_ui.castle-user-interface';
end;

procedure TTestCgeControl.FormCreate(Sender: TObject);
begin
  // Call this to have UI scaling, same as in editor
  CastleControl.Container.LoadSettings('castle-data:/CastleSettings.xml');

  CastleControl.Container.DesignUrl := 'castle-data:/test_3d.castle-user-interface';

  // adding a component created by code, doing manual rendering in TUiTest.Render
  CastleControl.Container.Controls.InsertFront(TUiTest.Create(Self));
end;

procedure TTestCgeControl.Timer1Timer(Sender: TObject);
begin
  LabelFps.Text := 'FPS: ' + CastleControl.Container.Fps.ToString;
end;

end.

File: examples/delphi/cpp_builder/fmx/Unit1.cpp
//---------------------------------------------------------------------------

#include <fmx.h>
#pragma hdrstop

#include "Unit1.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "Fmx.CastleControl"
// just test
//#pragma link "CastleScene"
//#pragma link "X3DNodes"
#pragma resource "*.fmx"
TForm1 *Form1;

// TUiTest --------------------------------------------------------------

/*
  Define simple TCastleUserInterface descendant, just to capture
  the moment OpenGL(ES) context is created and fill Memo1 will information.
*/
class TUiTest : public TCastleUserInterface
{
public:
	__fastcall TUiTest(TComponent* Owner) : TCastleUserInterface(Owner)
	{
	}

	virtual void __fastcall GLContextOpen()
	{
		Form1->Memo1->Lines->Add(GLInformationString());
	}
};

//---------------------------------------------------------------------------

__fastcall TForm1::TForm1(TComponent* Owner)
	: TForm(Owner)
{
	CastleControl1->Container->DesignUrl =
	  "castle-data:/test_3d.castle-user-interface";
	  //"castle-data:/test_2d.castle-user-interface";
	  //"castle-data:/test_ui.castle-user-interface";
	CastleControl1->Container->Controls()->InsertFront(new TUiTest(this));
}

//---------------------------------------------------------------------------
